### PR TITLE
Index configuration tool to create indices on a target cluster

### PR DESCRIPTION
**Pull request by kartg**
### Description
This folder (`index-configuration-tool`) houses Python code that automates the [pre-migration steps](https://github.com/kartg/opensearch-migrations/tree/datastash/datastash#2-configure-index-templates-on-the-target-cluster) of creating indices on the target cluster based on the settings and mappings in the source cluster.

The tool is capable of parsing a Logstash config file to retrieve endpoint information for source and target clusters. This includes validation to ensure that only Elasticsearch or Opensearch endpoints are supported. The parser grammar (in `logstash.lark`) is adapted from [node-logstash](https://github.com/bpaquet/node-logstash/blob/master/lib/logstash_config.jison) and uses [Lark](https://github.com/lark-parser/lark) for grammar definition, tree parsing and transformation (in the `logstash_conf_parser.py` module).

Once parsing and transformation is complete, the rest of the logic is housed in the `main.py` module itself. This includes:
* Querying both clusters to fetch index information and comparing them
* Printing a report of indices that conflict (and hence will not be recreated on the target cluster) and those that will be
* Creating indices on the target clusters with settings and mappings identical to the source cluster

All of these operations occur purely over HTTP - no ES/OpenSearch client libraries are used.

### Caveats

* This PR does not include tests. I've only tested this empirically and confirmed that it works correctly.
* The code has not been validated against edge cases and may not handle them well - HTTP connection failures, API differences, etc.
* The Logstash grammar definition does not yet support `if/else` statements
* Only [basic auth](https://opensearch.org/docs/latest/security/authentication-backends/authc-index/) is currently supported

### Check List
- [ ] New functionality includes testing
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: index-configuration-tool/main.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    for index in indices:
+        actual_endpoint = endpoint + index
+        data_dict = dict()
+        data_dict[SETTINGS_KEY] = indices[index][SETTINGS_KEY]
+        data_dict[MAPPINGS_KEY] = indices[index][MAPPINGS_KEY]
+        try:
+            resp = requests.put(actual_endpoint, auth=auth_tuple, json=data_dict)
+            resp.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            print("Failed to create index [" + index + "] - " + str(e), file=sys.stderr)
```
Comment (mikaylathompson) [inline index-configuration-tool/main.py:114]: minor, but f-string formatting is really nice for making this sort of thing easier to read:
```
print(f"Failed to create index [{index}] - {e}", file=sys.stderr)
```

**Comment by AmiStrn** [comment]
I admit i am a bit lost. what is the migration type? 
blue/green?

File: index-configuration-tool/logstash_conf_parser.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    plugin = tuple
+    STRING = var_name
+    PLUGIN_TYPE = var_name
+
+
+logstash_parser = Lark.open("logstash.lark", rel_to=__file__, parser="lalr", transformer=LogstashTransformer())
+
+
+def parse(logstash_file: str) -> dict:
+    conf_file = open(logstash_file, "r")
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/logstash_conf_parser.py:38]: Thank you! How obvious is it that i'm not super familiar with Python? 😛

File: index-configuration-tool/main.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+INTERNAL_SETTINGS_KEYS = ["creation_date", "uuid", "provided_name", "version", "store"]
+ALL_INDICES_ENDPOINT = "*"
+
+
+# Utility method to make a comma-separated string from a set
+def string_from_set(s: set) -> str:
+    if s:
+        return ", ".join(s)
+    else:
+        return "[]"
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/main.py:23]: Yes, the example above is correct. I'd like to communicate to users that a set is explicitly empty rather than printing an empty string, hence the `[]` return value. Since you noted that this is unusual, is there a better way to achieve this?

I'll update the type annotation to `set[str]` as you've suggested.

File: index-configuration-tool/main.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    endpoint += "/"
+    return endpoint, get_auth(plugin_config)
+
+
+def fetch_all_indices_by_plugin(plugin_config: dict) -> dict:
+    endpoint, auth_tuple = get_endpoint_info(plugin_config)
+    return fetch_all_indices(endpoint, auth_tuple)
+
+
+def fetch_all_indices(endpoint: str, auth_tuple: tuple) -> dict:
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/main.py:50]: > I assume requests.get handles a None value by just ignoring it?

Yes, that's correct

File: index-configuration-tool/main.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+# TODO Only supports basic auth for now
+def get_auth(input_data: dict) -> tuple:
+    if USER_KEY in input_data and PWD_KEY in input_data:
+        return input_data[USER_KEY], input_data[PWD_KEY]
+
+
+def get_endpoint_info(plugin_config: dict) -> tuple:
+    endpoint = "https://" if ("ssl" in plugin_config and plugin_config["ssl"]) else "http://"
+    endpoint += plugin_config[HOSTS_KEY][0] if type(plugin_config[HOSTS_KEY]) is list else plugin_config[HOSTS_KEY]
```
Comment (gregschohn) [inline index-configuration-tool/main.py:40]: When would this be a list?  Should you just take the first element or should you throw an error / make the user choose somehow.

File: index-configuration-tool/main.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+# Utility method to get the plugin config from the base data
+def get_plugin_config(data: dict, plugin_type: str) -> dict:
+    # Tuple contents are (type, config)
+    return data[plugin_type][1]
+
+
+# TODO Only supports basic auth for now
+def get_auth(input_data: dict) -> tuple:
+    if USER_KEY in input_data and PWD_KEY in input_data:
```
Comment (gregschohn) [inline index-configuration-tool/main.py:34]: wow - customers really write their credentials in plaintext into a config file?!

**Review by gregschohn — APPROVED** [review]
I'd like to see at least a quick and dirty test for the parse code & one for the diff() function too.  Other than that, minor comments - but I don't see anything that should block merge of this for now.

File: index-configuration-tool/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/main.py:40]: The `hosts` key for both the OS and [ES plugins](https://www.elastic.co/guide/en/logstash/7.10/plugins-inputs-elasticsearch.html#plugins-inputs-elasticsearch-hosts) could either be a simple string (in the case of a single host/endpoint) or an array of hosts (in which case Logstash round-robins through them per `bulk` request).

For this tool, we don't need to support round-robin since we're just creating indices and not ingesting data. Thus, the code here simply picks the first value if an array is provided. I've added the following comment to the code:

```
# "hosts" can be a simple string, or an array of hosts for Logstash to hit.
# This tool needs one accessible host, so we pick the first entry in the latter case.
```

**Comment by kartg** [comment]
> I admit i am a bit lost. what is the migration type?
> blue/green?

@AmiStrn B/G is our desired end state, but we're not quite there yet 😄 The purpose of this tool is to compare indices between a source and a target cluster and create the missing ones on the target cluster with the correct index settings and mappings. Once created, we would expect users to use Logstash (or [this recipe](https://github.com/kartg/opensearch-migrations/blob/datastash/datastash/README.md)) to move their existing data to the target cluster. The next step is to tie these two tools together (or integrate with [data-prepper](https://github.com/opensearch-project/data-prepper)) to allow users to easily migrate their ingested data across clusters. 

Once such "historical" data has been migrated, we have a reliable baseline to perform the rest of the blue/green migration. Tooling for this is still WIP, but will include:
* Dual writes to source and target cluster
* Shadow mode validation
* Cutover and rollback

Hopefully that clarifies things?

File: index-configuration-tool/main.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+MAPPINGS_KEY = "mappings"
+INDEX_KEY = "index"
+INTERNAL_SETTINGS_KEYS = ["creation_date", "uuid", "provided_name", "version", "store"]
+ALL_INDICES_ENDPOINT = "*"
+
+
+# Utility method to make a comma-separated string from a set
+def string_from_set(s: set) -> str:
+    if s:
+        return ", ".join(s)
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/main.py:21]: See https://github.com/opensearch-project/opensearch-migrations/pull/141#discussion_r1171712610 for the reasoning behind this

File: index-configuration-tool/main.py (hunk 6 above)
Comment (kartg) [reply index-configuration-tool/main.py:34]: Hence the name "basic" auth 😄 

I'll add support for more (read: better) auth mechanisms in the future

File: index-configuration-tool/main.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    identical_indices = set(indices_in_target) - set(index_conflicts)
+    indices_to_create = set(source.keys()) - set(indices_in_target)
+    # Print report
+    print("Indices in target cluster with conflicting settings/mappings: " + string_from_set(index_conflicts))
+    print("Indices already created in target cluster (data will NOT be moved): " + string_from_set(identical_indices))
+    print("Indices to create: " + string_from_set(indices_to_create))
+    return indices_to_create
+
+
+def create_indices(indices: dict, endpoint: str, auth_tuple: tuple):
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/main.py:104]: Renamed from simply `indices` to `indices_data`

File: index-configuration-tool/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikaylathompson) [reply index-configuration-tool/main.py:23]: Mm, okay, I see what you mean. I'm okay with the behavior -- it looks very asymmetric on its own, but the way it works in context seems reasonable.

**Review by mikaylathompson — APPROVED** [review]

File: index-configuration-tool/logstash.lark [hunk 9]
```diff
@@ -0,0 +1,38 @@
+// Adapated from https://github.com/bpaquet/node-logstash/blob/master/lib/logstash_config.jison
+// Tested via https://www.lark-parser.org/ide
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/logstash.lark:2]: Will work on adding tests next.

File: index-configuration-tool/logstash_conf_parser.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+from lark import Transformer
+
+
+class LogstashTransformer(Transformer):
+    def var_name(self, s):
+        return s.value
+
+    def string_literal(self, s):
+        (s,) = s
+        return s[1:-1]
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/logstash_conf_parser.py:14]: Good point. Parts of this code required some degree of trial-and-error, which is probably why it's not clear. The methods in the Transformer correspond to a `rule_name` or `TERMINAL` in the grammar definition. 

In the case of a `TERMINAL`, the method parameter is a [Token](https://lark-parser.readthedocs.io/en/latest/classes.html#token). In case of a `rule_name`, the method parameter is a `list` of its children in the AST. Note that the Transformer parses the AST bottom-up.

For now, i've added type hints and comments to make this clearer, but i think i can further improve the transformer to make it clearer. I'll work on this after implementing tests.

[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/logstash_conf_parser.py:14]: I've cleaned this up now so the Transformer doesn't operate on `TERMINAL` values, and uses only `rule_names`. I've also added comments.

File: index-configuration-tool/main.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    for index in result:
+        for setting in INTERNAL_SETTINGS_KEYS:
+            result[index][SETTINGS_KEY][INDEX_KEY].pop(setting, None)
+    return result
+
+
+def validate_logstash_config(config: dict):
+    if "input" not in config or "output" not in config:
+        raise ValueError("Missing input or output data from Logstash config")
+    for plugin_tuple in config["input"], config["output"]:
```
[the remark this answers is not in the document]
Comment (kartg) [reply index-configuration-tool/main.py:67]: This code has been refactored in https://github.com/opensearch-project/opensearch-migrations/pull/144 for better testability.
